### PR TITLE
Support span.type, prevent Datadog sampling traces

### PR DESCRIPTION
**Pull request by tomverran**
This PR does three things

- Maps a `span.type` tag into a `type` field on when we submit traces to Datadog
- Adds a magic metric that _should_ prevent Datadog only keeping a sample of our traces\*
- Adds an `X-DataDog-Trace-Count` header to stop the agent logging warnings all the time

\* This could be made configurable in the future but none of the services using this library to my knowledge have enough traffic to require us to drop some traces

### CODE REVIEW

File: natchez-datadog/src/main/scala/com/ovoenergy/effect/natchez/SubmittableSpan.scala [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+ * to submit traces. Turning a span into one of these involves some trickery
+ * with tags, hence this being its own file now
+ */
+case class SubmittableSpan(
+  traceId: Long,
+  spanId: Long,
+  name: String,
+  service: String,
+  resource: String,
+  `type`: Option[SpanType],
```
Comment (ericaovo) [inline natchez-datadog/src/main/scala/com/ovoenergy/effect/natchez/SubmittableSpan.scala:27]: how come this is an option?

File: natchez-datadog/src/main/scala/com/ovoenergy/effect/natchez/SubmittableSpan.scala [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+  implicit val encode: Encoder[SubmittableSpan] =
+    deriveConfiguredEncoder
+
+  implicit val decode: Decoder[SubmittableSpan] =
+    deriveConfiguredDecoder
+
+  /**
+   * It is very difficult to find any docs on this other than this fun Github issue by fommil
+   * https://github.com/DataDog/datadog-agent/issues/3031 but setting this magic metric to 2
+   * means the Datadog Agent should always keep all our traces.
```
Comment (ericaovo) [inline natchez-datadog/src/main/scala/com/ovoenergy/effect/natchez/SubmittableSpan.scala:56]: 🤞

File: natchez-datadog/src/main/scala/com/ovoenergy/effect/natchez/SubmittableSpan.scala [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+
+  /**
+   * Sadly the only API we have to set span type is through tags
+   * so we just have to guess if the user tried to set it
+   */
+  private def inferSpanType(tags: Map[String, TraceValue]): Option[SpanType] =
+    tags.collectFirst {
+      case ("span.type", StringValue("cache")) => SpanType.Cache
+      case ("span.type", StringValue("web")) => SpanType.Web
+      case ("span.type", StringValue("db")) => SpanType.Db
```
Comment (ericaovo) [inline natchez-datadog/src/main/scala/com/ovoenergy/effect/natchez/SubmittableSpan.scala:89]: what about `custom`?

File: natchez-datadog/src/main/scala/com/ovoenergy/effect/natchez/SubmittableSpan.scala (hunk 1 above)
Comment (tomverran) [reply natchez-datadog/src/main/scala/com/ovoenergy/effect/natchez/SubmittableSpan.scala:27]: Any spans we don't explicitly add a type to become `Custom` - which is also why I don't turn the `Custom` span type into a string when submitting spans to Datadog

File: natchez-datadog/src/main/scala/com/ovoenergy/effect/natchez/SubmittableSpan.scala (hunk 3 above)
Comment (tomverran) [reply natchez-datadog/src/main/scala/com/ovoenergy/effect/natchez/SubmittableSpan.scala:89]: replied above :+1:  - a `None` implies `custom`

**Comment by ericaovo** [comment]
what are you referring to with `sample of our traces`?

**Comment by tomverran** [comment]
If you have an app with lots of traffic you can ask the Datadog agent to only submit a sample of the traces it recieves up to Datadog: https://docs.datadoghq.com/tracing/guide/trace_sampling_and_storage/?tab=java

File: natchez-datadog/src/main/scala/com/ovoenergy/effect/natchez/DatadogTags.scala [hunk 4]
```diff
@@ -11,6 +11,8 @@ object DatadogTags {
   sealed trait SpanType
 
   object SpanType {
+    case object Custom extends SpanType
+    case object Cache extends SpanType
```
Comment (mwz) [inline natchez-datadog/src/main/scala/com/ovoenergy/effect/natchez/DatadogTags.scala:15]: I can see this type in the docs, but I'm not sure what it means. Any idea what it's supposed to be used for?

Comment (tomverran) [reply natchez-datadog/src/main/scala/com/ovoenergy/effect/natchez/DatadogTags.scala:15]: I don't really know but my assumption is it is for profiling something like a Redis or Memcached cache

Comment (mwz) [reply natchez-datadog/src/main/scala/com/ovoenergy/effect/natchez/DatadogTags.scala:15]: yeah, does what it says on the tin =)

**Pull request merged**